golang/thrift/inbound.go

```diff
@@ -0,0 +1,297 @@
+package thrift
+
+import "bytes"
+import "errors"
+import "fmt"
+import "strings"
+import "github.com/apache/thrift/lib/go/thrift"
+import tchannel "github.com/uber/tchannel/golang"
+
+// NewTChannelInboundProtocol creates a TChannelInboundProtocol
+func NewTChannelInboundProtocol(call *tchannel.InboundCall) *TChannelInboundProtocol {
+	return &TChannelInboundProtocol{
+		call: call,
+	}
+}
+
+//
+// TChannelInboundProtocol is a thrift.TProtocol implementation for
+// inbound (i.e., server side) tchannel calls. It serves as the adaptor
+// between the generated thrift server side code (thrift.TProcessor)
+// and tchannel (tchannel.InboundCall).
+//
+// Incoming and outgoing data is buffered in memory buffers and the
+// actual parsing of the data is delegated to thrift.TBinaryProtocol.
+//
+type TChannelInboundProtocol struct {
+	call *tchannel.InboundCall
+
+	writer      *thrift.TBinaryProtocol
+	writeBuffer *MemoryBufferTransport
```

Comment:
This double buffering is a shame.  We should probable extend InboundCall / InboundCallResponse with BeginArg3 methods that return an io.Reader/io.Writer, so that the thrift.TBinaryProtocol to stream directly to them vs needing to go through an additional intermediate in-memory buffer.